src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -0,0 +1,144 @@
+                if (!_knownAssemblyPathsBySimpleName.TryGetValue(simpleName, out paths))
+                {
+                    _knownAssemblyPathsBySimpleName.Add(simpleName, new List<string>() { fullPath });
+                }
+                else if (!paths.Contains(fullPath))
+                {
+                    paths.Add(fullPath);
+                }
+            }
+        }
+
+        public Assembly LoadFromPath(string fullPath)
+        {
+            CompilerPathUtilities.RequireAbsolutePath(fullPath, nameof(fullPath));
+            return LoadFromPathUnchecked(fullPath);
+        }
+
+        #endregion
+
+        private Assembly LoadFromPathUnchecked(string fullPath)
+        {
+            Debug.Assert(PathUtilities.IsAbsolute(fullPath));
+
+            lock (_guard)
+            {
+                Assembly existingAssembly;
+                if (_loadedAssembliesByPath.TryGetValue(fullPath, out existingAssembly))
+                {
+                    return existingAssembly;
+                }
+            }
+
+            Assembly assembly = LoadFromPathImpl(fullPath);
+            return AddToCache(assembly, AssemblyIdentity.FromAssemblyDefinition(assembly), fullPath);
+        }
+
+        public Assembly Load(string displayName)
+        {
+            AssemblyIdentity requestedIdentity;
+            if (!AssemblyIdentity.TryParseDisplayName(displayName, out requestedIdentity))
+            {
+                return null;
+            }
+
+            ImmutableArray<string> candidatePaths;
+            lock (_guard)
+            {
+                Assembly existingAssembly;
+                
+                // First, check if this loader already loaded the requested assembly:
+                if (_loadedAssembliesByIdentity.TryGetValue(requestedIdentity, out existingAssembly))
+                {
+                    return existingAssembly;
+                }
+
+                // Second, check if an assembly file of the same simple name was registered with the loader:
+                List<string> pathList;
+                if (!_knownAssemblyPathsBySimpleName.TryGetValue(requestedIdentity.Name, out pathList))
+                {
+                    return null;
+                }
+
+                Debug.Assert(pathList.Count > 0);
+                candidatePaths = pathList.ToImmutableArray();
+            }
+
+            // Multiple assemblies of the same simple name but different identiries might have been registered.
+            // Load the one that matches the requested identity (if any).
+            foreach (var candidatePath in candidatePaths)
+            {
+                var candidateIdentity = AssemblyIdentityUtils.TryGetAssemblyIdentity(candidatePath);
+
+                if (requestedIdentity.Equals(candidateIdentity))
+                {
+                    var assembly = LoadFromPathUnchecked(candidatePath);
+                    return AddToCache(assembly, requestedIdentity, candidatePath);
```

Comment:
AddToCache seems unnecessary here - LoadFromPathUnchecked already does the caching.